crux-core/src/crux/kv/index_store.clj

```diff
@@ -535,6 +536,35 @@
                     (new-prefix-kv-iterator tx-time-mapping-prefix))]
     (latest-completed-tx-i i)))
 
+(defprotocol PThreadManager
+  (snapshot-opened [_ snapshot])
+  (snapshot-closed [_ snapshot]))
+
+(defrecord ThreadManager [^Map snapshot-threads]
+  PThreadManager
+  (snapshot-opened [this snapshot]
+    (locking this
+      (.put snapshot-threads snapshot (Thread/currentThread))))
+
+  (snapshot-closed [this snapshot]
+    (locking this
+      (.remove snapshot-threads snapshot)
+      (.notifyAll this)))
```

Comment:
I would also recommend not using `notifyAll` / `wait` directly, but some higher level `java.util.concurrent.`. construct depending on what you trying to do, would also remove the need for explicit `locking` most likely. In this case you might not need anything else, but just use a simple polling loop in close.

Reply:
migrated to use a Semaphore